File: python/nucleotide-count/dna.py
from collections import Counter

class DNA:

  COUNTABLE_NUCLEOTIDES = "ACGTU"

  def __init__(self, sequence):
    self.counter = Counter(sequence)

  def count(self, nucleotide):
    if nucleotide in DNA.COUNTABLE_NUCLEOTIDES:
      return self.counter[nucleotide]
    else:
      raise ValueError, ("%s is not a nucleotide." % nucleotide)

  def nucleotide_counts(self):
    return dict(self._default_nucleotide_counts() + self.counter.items())

  def _default_nucleotide_counts(self):
    return {"A": 0, "C": 0, "G": 0, "T": 0}.items()

File: clojure/nucleotide-count/dna.clj
(ns dna)
(require '[clojure.string :as string])
(use '[clojure.set :only [union]])

(def common-nucleotides #{"A" "C" "G"})
(def dna-nucleotides (conj common-nucleotides "T"))
(def rna-nucleotides (conj common-nucleotides "U"))
(def countable-nucleotides (union dna-nucleotides rna-nucleotides))
(def default-nucleotide-counts (reduce #(assoc %1 %2 0) {} dna-nucleotides))

(defn- nucleotides [sequence]
  (map str sequence))

(defn- countable-nucleotide? [nucleotide]
  (contains? countable-nucleotides nucleotide))

(defn- nucleotide-counter [nucleotide]
  #(if (= %2 nucleotide) (inc %1) %1))

(defn count [nucleotide sequence]
  (if (countable-nucleotide? nucleotide)
    (reduce (nucleotide-counter nucleotide) 0 (nucleotides sequence))
    (throw (Exception. "invalid nucleotide"))))

(defn nucleotide-counts [sequence]
  (merge default-nucleotide-counts (frequencies (nucleotides sequence))))